### PR TITLE
Aviad/blake2s

### PR DESCRIPTION
**Pull request by aviadingo**
## Describe the changes

adds Blake2s cuda capability.

### CODE REVIEW

File: icicle/src/hash/blake2s/blake2s.cu [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    {6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5}, {10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0}};
+
+  __device__ uint32_t cuda_blake2s_leuint32(const BYTE* in)
+  {
+    uint32_t a;
+    memcpy(&a, in, 4);
+    return a;
+  }
+
+  __device__ uint32_t cuda_blake2s_ROTR32(uint32_t a, uint8_t b) { return (a >> b) | (a << (32 - b)); }
```
Comment (ChickenLover) [inline icicle/src/hash/blake2s/blake2s.cu:75]: Maybe worth to add `__inline__`

File: icicle/src/hash/blake2s/blake2s.cu [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  mcm_cuda_blake2s_hash_batch(BYTE* key, WORD keylen, BYTE* in, WORD inlen, BYTE* out, WORD output_len, WORD n_batch)
+  {
+    BYTE* cuda_indata;
+    BYTE* cuda_outdata;
+    const WORD BLAKE2S_BLOCK_SIZE = output_len;
+    cudaMalloc(&cuda_indata, inlen * n_batch);
+    cudaMalloc(&cuda_outdata, BLAKE2S_BLOCK_SIZE * n_batch);
+    assert(keylen <= 32);
+
+    // CUDA_BLAKE2S_CTX ctx;
```
Comment (ChickenLover) [inline icicle/src/hash/blake2s/blake2s.cu:241]: These should be removed

File: icicle/src/hash/blake2s/blake2s.cu [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    // cpu_blake2s_init(&ctx, key, keylen, n_outbit);
+    // cudaMemcpyToSymbol(c_CTX, &ctx, sizeof(CUDA_BLAKE2S_CTX), 0, cudaMemcpyHostToDevice);
+
+    cudaMemcpy(cuda_indata, in, inlen * n_batch, cudaMemcpyHostToDevice);
+
+    WORD thread = 256;
+    WORD block = (n_batch + thread - 1) / thread;
+    kernel_blake2s_hash<<<block, thread>>>(cuda_indata, inlen, cuda_outdata, n_batch, BLAKE2S_BLOCK_SIZE);
+    cudaMemcpy(out, cuda_outdata, BLAKE2S_BLOCK_SIZE * n_batch, cudaMemcpyDeviceToHost);
+    cudaDeviceSynchronize();
```
Comment (ChickenLover) [inline icicle/src/hash/blake2s/blake2s.cu:251]: Current implementation does not support async. All of our other primitives do. So maybe worth adding HashConfig as an input and changing all the functions to their async alternatives

File: icicle/src/hash/blake2s/blake2s.cu [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    cudaError_t error = cudaGetLastError();
```
Comment (ChickenLover) [inline icicle/src/hash/blake2s/blake2s.cu:252]: Please use our error-management functions (you can find an example in any of our primitives)

**Review by ChickenLover — COMMENTED** [review]
Good job with the PR. I left a bunch of style related comments. In theory they can be ignored as we are merging this to V2

File: icicle/src/hash/blake2s/extern.cu [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "hash/blake2s/blake2s.cuh"
+#include "blake2s.cu"
+#include "../../merkle-tree/merkle.cu"
+#include "merkle-tree/merkle.cuh"
+
+namespace blake2s {
+  extern "C" cudaError_t blake2s_cuda(
+    BYTE* input, BYTE* output, WORD number_of_blocks, WORD input_block_size, WORD output_block_size, HashConfig& config)
+  {
```
Comment (emirsoyturk) [inline icicle/src/hash/blake2s/extern.cu:14]: I noticed that the parameter order in `blake2s_cuda` is a bit different from `keccak256_cuda` (and other keccak functions). 

```cpp
keccak256_cuda(uint8_t* input, int input_block_size, int number_of_blocks, uint8_t* output, HashConfig& config)
```

To keep things consistent, so my suggestion smt like this:


```cpp
blake2s_cuda(BYTE* input, WORD input_block_size, WORD number_of_blocks, WORD output_block_size, BYTE* output, HashConfig& config)
```

**Review by jeremyfelder — APPROVED** [review]

**Review by emirsoyturk — APPROVED** [review]

**Review by mickeyasa — COMMENTED** [review]
do we need this PR?

**Comment by jeremyfelder** [comment]
We have a separate branch for the stwo integration which this has been integrated into so we can close the PR

**Pull request closed without merging**